sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java

```diff
@@ -334,8 +335,14 @@ public void setCoderRegistry(CoderRegistry coderRegistry) {
   @Internal
   public interface PipelineVisitor {
     /**
-     * Called for each composite transform after all topological predecessors have been visited
-     * but before any of its component transforms.
+     * Called before visiting anything values or transforms, as many uses of a visitor require
+     * access to the {@link Pipeline} object itself.
+     */
+    void visitPipeline(Pipeline p);
```

Comment:
Perhaps this can be `enterPipeline/exitPipeline`, and we can get rid of the root node (and bring the `TransformHierarchy` in line with the runner API `Pipeline` definition)